src/compiler/sys.ts

```diff
@@ -869,6 +875,63 @@ namespace ts {
             };
             return nodeSystem;
 
+            /**
+             * Uses the builtin inspector APIs to capture a CPU profile
+             * See https://nodejs.org/api/inspector.html#inspector_example_usage for details
+             */
+            function enableCPUProfiler(path: string, cb: () => void) {
+                if (activeSession) {
+                    cb();
+                    return false;
+                }
+                const inspector: typeof import("inspector") = require("inspector");
+                const session = new inspector.Session();
+                session.connect();
+
+                session.post("Profiler.enable", () => {
+                    session.post("Profiler.start", () => {
+                        activeSession = session;
+                        profilePath = path;
+                        cb();
+                    });
+                });
+                return true;
+            }
+
+            function disableCPUProfiler(cb: () => void) {
+                if (activeSession && activeSession !== "stopping") {
+                    const s = activeSession;
+                    activeSession.post("Profiler.stop", (err, { profile }) => {
+                        if (!err) {
+                            try {
+                                if (_fs.statSync(profilePath).isDirectory()) {
+                                    profilePath = _path.join(profilePath, `${(new Date()).toISOString().replace(/:/g, "-")}+P${process.pid}.cpuprofile`);
+                                }
+                            }
+                            catch {
+                                // do nothing and ignore fallible fs operation
+                            }
+                            try {
+                                _fs.mkdirSync(_path.dirname(profilePath), { recursive: true });
+                            }
+                            catch {
+                                // do nothing and ignore fallible fs operation
+                            }
+                            _fs.writeFileSync(profilePath, JSON.stringify(profile));

```

Comment:
If the goal of this is to allow users to send us a CPU profile to investigate an issue, we should ensure no sensitive data can be captured as part of the profile they would send. Should we have a pass over the CPU profile to replace possibly sensitive path data?

Reply:
It'd just be the name of the path to tsc, no?

Reply:
Maybe we could shorten all paths to just filenames?

Reply:
> Maybe we could shorten all paths to just filenames?

Soo... filenames are OK but full paths aren't? While in the case of `tsc.js` (which is all I'm implementing here), all the (not node-internal) paths in the profile are to `tsc.js`, in the LS/API case in theory someone could use this to capture a trace when the LS has been wrapped by a host and some frames come from someone else's code~

Reply:
Our offline conversation had led me to believe that all filenames would be from our source.  If that's not the case, we might prefer to do additional sanitization.

Reply:
Filenames would generally be ours, but we could be installed under any path:
`c:\dev\topsecretproject\node_modules\lib\tsc.js` would be bad.

Here's a rough outline for a CpuProfile path sanitizer (untested):
```ts
const nativePattern = /^native |^\([^)]+\)$|^(internal[\\/]|[a-zA-Z0-9_\s]+(\.js)?$)/;
function cleanupPaths(profile: import("inspector").Profiler.Profile) {
    let externalFileCounter = 0;
    for (const node of profile.nodes) {
        if (node.callFrame.url) {
            const url = normalizeSlashes(node.callFrame.url);
            if (containsPath(__dirname, url, useCaseSensitiveFileNames)) {
                node.callFrame.url = convertToRelativePath(url, __dirname, createGetCanonicalFileName(useCaseSensitiveFileNames));
            }
            else if (!nativePattern.test(url)) {
                node.callFrame.url = `external${externalFileCounter}.js`;
                externalFileCounter++;
            }
        }
    }
}
```

It strips the path prefix off of typescript paths, retains paths for v8 natives and node internals, and replaces all other urls with `external${n++}.js`.

Reply:
Are we sure we should be sanitizing that stuff in our own code, and not in _whatever_ automated thing may in the future ingest these from LS operations? The `sysLog` logging function spews paths to those diagnostic logs, and we don't filter those...

Reply:
Like.... this seems... premature.... these dumps don't contain any more data than the tsserver logs already do... less, on the whole, even, and I assume we already have a pipeline for sanitizing those on any editor-based flow that automatically submits them.

Reply:
*sigh* It's done, but I kinda feel like this kind of sanitization shouldn't be in the tool itself, and should instead be in some kind of ingestion mechanism, but w/e.